pennylane/ops/mid_circuit_measure.py

```diff
@@ -0,0 +1,291 @@
+
+    def __rmul__(self, other: Any):
+        return apply_to_measurement_dependant_values(lambda x, y: y * x)(self, other)
+
+    def __str__(self):
+        measurements = self.measurements
+        lines = []
+        for k, v in self.branches.items():
+            lines.append(
+                ",".join([f"{measurements[i]}={k[i]}" for i in range(len(measurements))])
+                + " => "
+                + str(v)
+            )
+        return "\n".join(lines)
+
+    def _merge(self, other: Union["MeasurementDependantValue", "_Value", Any]):
+        """
+        Merge this MeasurementDependantValue with `other`.
+
+        Ex: Merging a MeasurementDependantValue such as
+
+        .. code-block:: python
+
+            df3jff4t=0 => 3.4
+            df3jff4t=1 => 1
+
+        with another MeasurementDependantValue:
+
+        .. code-block:: python
+
+            f93fjdj3=0 => 100
+            f93fjdj3=1 => 67
+
+        will result in:
+
+        .. code-block:: python
+
+            df3jff4t=0,f93fjdj3=0 => 3.4,100
+            df3jff4t=0,f93fjdj3=1 => 3.4,67
+            df3jff4t=1,f93fjdj3=0 => 1,100
+            df3jff4t=1,f93fjdj3=1 => 1,67
+
+        (note the uuids in the example represent distinct measurements of different qubit.)
+
+        """
+        if isinstance(other, MeasurementDependantValue):
+            if self._dependent_on == other._dependent_on:
+                return MeasurementDependantValue(
+                    self._dependent_on,
+                    self._zero_case._merge(other._zero_case),
+                    self._one_case._merge(other._one_case),
+                )
+            if self._dependent_on < other._dependent_on:
+                return MeasurementDependantValue(
+                    self._dependent_on,
+                    self._zero_case._merge(other),
+                    self._one_case._merge(other),
+                )
+            return MeasurementDependantValue(
+                other._dependent_on,
+                self._merge(other._zero_case),
+                self._merge(other._one_case),
+            )
+        return MeasurementDependantValue(
+            self._dependent_on,
+            self._zero_case._merge(other),
+            self._one_case._merge(other),
+        )
+
+    def _transform_leaves_inplace(self, fun: Callable):
+        """
+        Transform the leaves of a MeasurementDependantValue with `fun`.
+        """
+        self._zero_case._transform_leaves_inplace(fun)
+        self._one_case._transform_leaves_inplace(fun)
+
+
+# pylint: disable=too-few-public-methods,protected-access
+class _Value(Generic[T]):
+    """
+    Leaf node for a MeasurementDependantValue tree structure.
+    """
+
+    __slots__ = ("_values",)
+
+    def __init__(self, *values):
+        self._values = values
+
+    def _merge(self, other):
+        """
+        Works with MeasurementDependantValue._merge
```

Comment:
it just means that, `_merge` for `MeasurementDependentValue` is a recursive algorithm, and as the calls bubble down, it will eventually call to this `_merge`.